src/v/cluster/archival/tests/archiver_scheduler_impl_gtest.cc

```diff
@@ -0,0 +1,346 @@
+    //   limit of 300ms
+    std::chrono::milliseconds initial_backoff = 100ms;
+    std::chrono::milliseconds max_backoff = 300ms;
+    model::ntp expected_ntp(
+      model::kafka_namespace,
+      model::topic("panda-topic"),
+      model::partition_id(137));
+    ss::abort_source as;
+    basic_retry_chain_node<ss::manual_clock> rtcnode(as);
+    scoped_config cfg;
+    cfg.get("cloud_storage_upload_loop_initial_backoff_ms")
+      .set_value(initial_backoff);
+    cfg.get("cloud_storage_upload_loop_max_backoff_ms").set_value(max_backoff);
+
+    archiver_scheduler<ss::manual_clock> scheduler(100, 10);
+
+    co_await scheduler.start();
+
+    co_await scheduler.create_ntp_state(expected_ntp);
+
+    // First call
+    auto fut1 = scheduler.maybe_suspend_upload(
+      upload_resource_usage<ss::manual_clock>{
+        .ntp = expected_ntp,
+        .put_requests_used = 2,
+        .uploaded_bytes = 20,
+        .errc = error_outcome::not_enough_data,
+        .archiver_rtc = std::ref(rtcnode),
+      });
+    ASSERT_FALSE_CORO(fut1.available());
+
+    ss::manual_clock::advance(initial_backoff - 1ms);
+    ASSERT_FALSE_CORO(fut1.available());
+
+    ss::manual_clock::advance(3ms);
+    auto quota1 = co_await std::move(fut1);
+
+    ASSERT_TRUE_CORO(quota1.has_value());
+    ASSERT_EQ_CORO(quota1.value().upload_size_quota, 80);
+    ASSERT_EQ_CORO(quota1.value().requests_quota, 8);
+
+    // second call
+    auto fut2 = scheduler.maybe_suspend_upload(
+      upload_resource_usage<ss::manual_clock>{
+        .ntp = expected_ntp,
+        .put_requests_used = 3,
+        .uploaded_bytes = 30,
+        .errc = error_outcome::not_enough_data,
+        .archiver_rtc = std::ref(rtcnode),
+      });
+    ASSERT_FALSE_CORO(fut2.available());
+
+    ss::manual_clock::advance(initial_backoff * 2 - 1ms);
+    ASSERT_FALSE_CORO(fut2.available());
+
+    ss::manual_clock::advance(3ms);
+    auto quota2 = co_await std::move(fut2);
+
+    ASSERT_TRUE_CORO(quota2.has_value());
+    // Some units will be returned
+    ASSERT_GT_CORO(quota2.value().upload_size_quota, 50);
+    ASSERT_GT_CORO(quota2.value().requests_quota, 5);
+
+    // third call
+    // at this stage we should reach max backoff interval of 300ms
+    auto fut3 = scheduler.maybe_suspend_upload(
+      upload_resource_usage<ss::manual_clock>{
+        .ntp = expected_ntp,
+        .put_requests_used = 4,
+        .uploaded_bytes = 40,
+        .errc = error_outcome::not_enough_data,
+        .archiver_rtc = std::ref(rtcnode),
+      });
+    ASSERT_FALSE_CORO(fut3.available());
+
+    ss::manual_clock::advance(initial_backoff * 3 - 1ms);
+    ASSERT_FALSE_CORO(fut3.available());
+
+    ss::manual_clock::advance(3ms);
+    auto quota3 = co_await std::move(fut3);
+
+    ASSERT_TRUE_CORO(quota3.has_value());
+    // Some units will be returned
+    ASSERT_GT_CORO(quota3.value().upload_size_quota, 10);
+    ASSERT_GT_CORO(quota3.value().requests_quota, 1);
```

Comment:
Does it make sense to maybe_suspend_upload() after this with a success case followed by another not_enough_data case this to check that backoffs reset?